plugins/analysis-icu/src/main/java/org/elasticsearch/index/mapper/ICUCollationFieldMapper.java

```diff
@@ -0,0 +1,623 @@
+                        iterator.remove();
+                        break;
+                    case "variant":
+                        builder.variant(XContentMapValues.nodeStringValue(fieldNode, null));
+                        iterator.remove();
+                        break;
+                    case "strength":
+                        builder.strength(XContentMapValues.nodeStringValue(fieldNode, null));
+                        iterator.remove();
+                        break;
+                    case "decomposition":
+                        builder.decomposition(XContentMapValues.nodeStringValue(fieldNode, null));
+                        iterator.remove();
+                        break;
+                    case "alternate":
+                        builder.alternate(XContentMapValues.nodeStringValue(fieldNode, null));
+                        iterator.remove();
+                        break;
+                    case "caseLevel":
+                        builder.caseLevel(XContentMapValues.nodeBooleanValue(fieldNode, false));
+                        iterator.remove();
+                        break;
+                    case "caseFirst":
+                        builder.caseFirst(XContentMapValues.nodeStringValue(fieldNode, null));
+                        iterator.remove();
+                        break;
+                    case "numeric":
+                        builder.numeric(XContentMapValues.nodeBooleanValue(fieldNode, false));
+                        iterator.remove();
+                        break;
+                    case "variableTop":
+                        builder.variableTop(XContentMapValues.nodeStringValue(fieldNode, null));
+                        iterator.remove();
+                        break;
+                    case "hiraganaQuaternaryMode":
+                        builder.hiraganaQuaternaryMode(XContentMapValues.nodeBooleanValue(fieldNode, false));
+                        iterator.remove();
+                        break;
+                    default:
+                        break;
+                }
+            }
+
+            TypeParsers.parseField(builder, name, node, parserContext);
+            return builder;
+        }
+    }
+
+    private final String rules;
+    private final String language;
+    private final String country;
+    private final String variant;
+    private final String strength;
+    private final String decomposition;
+    private final String alternate;
+    private final boolean caseLevel;
+    private final String caseFirst;
+    private final boolean numeric;
+    private final String variableTop;
+    private final boolean hiraganaQuaternaryMode;
+    private final Collator collator;
+
+    // each thread needs its own ICUCollationDocValuesField because the internally cloned collator is not thread-safe
+    private final ThreadLocal<ICUCollationDocValuesField> perThreadICUCollationField;
```

Comment:
I did this because of [this comment](https://github.com/apache/lucene-solr/blob/master/lucene/analysis/icu/src/java/org/apache/lucene/collation/ICUCollationDocValuesField.java#L45).  Not sure how expensive cloning a `Collator` is but it will happen for every value being indexed if we create them on demand right?